src/test/ui/stability-attribute/stability-attribute-trait-impl.rs

```diff
@@ -23,6 +23,13 @@ impl StableTrait for UnstableType {}
 impl UnstableTrait for StableType {}
 
 #[unstable(feature = "h", issue = "none")]
+impl StableTrait for ! {}
+
+// Note: If C-unwind is stabilized, switch this to another (unstable) ABI.
+#[unstable(feature = "i", issue = "none")]
+impl StableTrait for extern "C-unwind" fn() {}
+
```

Comment:
```suggestion

#[unstable(feature = "i", issue = "none")]
//~^ ERROR an `#[unstable]` annotation here has no effect [ineffective_unstable_trait_impl]
impl StableTrait for fn() -> ! {}
```

Reply:
This test actually already passes, because i don't recurse into the return type in the BareFn case.

But that also means it thinks `fn() -> UnstableType` is stable. Not sure if it's worth fixing that, since that just means it is extra conservative with unstably implementing stable traits for such types.

Reply:
Fixed. It now understands both `fn() -> !` and `fn() -> UnstableType`.